// smartcontract-ui-fn/parameter-form.js
const bel = require('bel')
const csjs = require('csjs-inject')

const sub = require('subprotocols')

const inputPayable = require("input-payable")
const inputArray = require("input-array")
const inputInteger = require("input-integer")
const inputByte = require("input-byte")
const inputString = require("input-string")
const inputBoolean = require("input-boolean")
const inputAddress = require("input-address")

const id = 'parameter-form'
var counter = 1

module.exports = makeParameterForm

function makeParameterForm ({ data, theme = {} }, protocol) {
  const name = `${id}/${counter++}`
  const notify = protocol(msg => console.log(`[${name}] receives:`, msg))

  const css = Object.assign({}, classes, theme.classes)
  const vars = Object.assign({}, variables, theme.variables)

  console.log({protocol, sub})
  console.error('@TODO: deal with parent `opts` and global `theme`')
  console.error('@TODO: connect `protocol` with `_protocol`')
  const _protocol = sub('parameter-form', (...args) => console.log('custoomize parameter form', args))
  debugger

  const output = bel`<span class=${css.output}></span>`

  return data.map(x => {
    if (x.components) return bel`<li>
      <div>${x.name} (${x.type})</div>
      <ul>${makeParameterForm(x.components)}</ul>
    </li>`
    if (!x.components) return generateInputContainer(x)
  })
  function generateInputContainer (field) {
    var theme = { colors: vars }
    var name = field.name
    var type = field.type
    if (type === 'constructor') alert('foo')
    var inputField = getInputField({ theme, type }, _protocol)
    var inputContainer = bel`
      <div class=${css.inputContainer}>
        <label class=${css.inputParam} title="data type: ${type}">${name || 'key'}</label>
        <div class=${css.inputFields}>${inputField}</div>
      </div>`
    return inputContainer
    function cb (msg, el, value) {
      output.innerHTML = ""
      output.innerHTML = msg ? `<span class=${css.valError} title="${msg}">
        <i class="fa fa-exclamation"></i>
      </span>` : `<span class=${css.valSuccess} title="The value is valid.">
        <i class="fa fa-check"></i>
      </span>`
      el.parentNode.appendChild(output)
    }
  }
  function getInputField ({ theme, type }, protocol) {
    var field
    if ((type.search(/\]/) != -1)) field = inputArray({ theme, type }, protocol.sub('input-array'))
    else {
      if ((type.search(/\buint/) != -1) || (type.search(/\bint/) != -1)) field = inputInteger({ theme, type }, protocol.sub('input-uint'))
      if (type.search(/\bbyte/) != -1) field = inputByte({ theme, type }, protocol.sub('input-byte'))
      if (type.search(/\bstring/) != -1) field = inputString({ theme, type }, protocol.sub('input-string'))
      // if (type.search(/\bfixed/) != -1) field = inputInteger({ theme, type }, protocol.sub('input-fixed'))
      if (type.search(/\bbool/) != -1) field = inputBoolean({ theme, type }, protocol)
      // function inputProtocol (notify) {
      //   return msg => {
      //     console.log(`[${name}] receives:`, msg)
      //     const { from, type, body } = msg
      //     if (type === 'data') notify({ from: [name, ...from], type, body })
      //   }
      // }
      if (type.search(/\baddress/) != -1) field = inputAddress({ theme, type }, protocol.sub('input-'))
    }
    return field
  }

}
const variables = { // default values
  inputParamPadding: '',
  inputParamColor: '',
  inputParamFontSize: '',
  inputParamTextAlign: '',
  valErrorColor: '',
  valErrorBackgroundColor: '',
  valSuccessColor: '',
  valSuccessBackgroundColor: '',
}
const classes = csjs`
.inputContainer {
  display: grid;
  grid-template-columns: 25% auto;
  grid-template-rows: auto;
  margin-bottom: 22px;
  position: relative;
  z-index: 3;
}
.inputParam {
  padding: var(--inputParamPadding);
  color: var(--inputParamColor);
  font-size: var(--inputParamFontSize);
  text-align: var(--inputParamTextAlign);
  word-break: break-all;
  transition: color .3s ease-in-out;
}
.inputFields {
  position: relative;
  display: inline-grid;
  grid-row-gap: 20px;
  align-items: center;
}
.output {
  position: absolute;
  top: 5px;
  right: -25px;
  align-self: center;
}
.output i {
  font-size: 1.2rem;
}
.output span {
  display: inline-block;
  border-radius: 30px;
  width: 20px;
  height: 20px;
  text-align: center;
}
.valError {
  color: var(--valErrorColor);
  background-color: var(--valErrorBackgroundColor);
  margin-top: 2px;
  transition: colors .6s, background-color .6s ease-in-out;
}
.valSuccess {
  color: var(--valSuccessColor);
  background-color: var(--valSuccessBackgroundColor);
  transition: colors .6s, background-color .6s ease-in-out;
}`
